File: learn-algo/leetcode/1995/c++/1995.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <numeric>
#include <math.h>
#include <set>
#include <map>
#include <unordered_map>
#include <unordered_set>
#include <queue>

using namespace std;

class Solution {
public:
    int countQuadruplets(vector<int> &nums) {
        int n = nums.size();
        int ans = 0;
        unordered_map<int, int> tMap;
        for (int b = n - 3; b >= 1; b--) {
            for (int d = n - 1; d > b + 1; d--) {
                int c = b + 1;
                int t = nums[d] - nums[c];
                tMap[t]++;
            }

            for (int a = 0; a < b; a++) {
                ans += tMap[nums[a] + nums[b]];
            }
        }

        return ans;
    }
};

int main() {
    cout << endl;
    return 0;
}

File: learn-algo/leetcode/1995/go/1995.go
package main

func countQuadruplets(nums []int) int {
	n := len(nums)
	tMap := map[int]int{}
	ans := 0

	for c := n - 2; c >= 2; c-- {
		tMap[nums[c+1]]++
		for a := 0; a < c-1; a++ {
			for b := a + 1; b < c; b++ {
				t := nums[a] + nums[b]
				ans += tMap[t]
			}
		}
	}

	return ans
}

func countQuadruplets2(nums []int) int {
	n := len(nums)
	tMap := map[int]int{}
	ans := 0

	for b := n - 3; b >= 1; b-- {
		for d := n - 1; d > b+1; d-- {
			tMap[nums[d]-nums[b+1]]++
		}

		for a := 0; a < b; a++ {
			ans += tMap[nums[a]+nums[b]]
		}
	}
	return ans
}
